kobotoolbox/kpi
# Faster frontend builds

## Description

Improve frontend build speed.

## Notes for reviewer

Main goal of this PR:

- Make the dev server start faster and consume less resources while running.

While not the main goal, I took the opportunity to modify other areas of the build system, including:

- Speed up prod builds.
- Speed up test builds.
- Auto-rebuild tests on change and run them in a browser tab
- CI tests - labeled steps, granular, and more reliable.
- Storybook builds, a bit faster (not amazing though)
- I made it a little clearer how to run linters individually, by adding "scripts" entries in package.json.

And caught a few things during cleanup:

- Stop shipping an extra copy of polyfills for browsers we no longer support
- Stop putting extra '.' when processing frontend assets, e.g. `image..gif` -> `image.gif`

### Main result

- `3x-5x` faster start time for dev server (depends on the machine)
- Similar improvements to prod build speed
    - from ~`2m`  to ~`30s`
- Linters are moved out of the core webpack build, but ESLint still runs in CI.
- Stylelint and CoffeeLint now have entries in package.json to document how to run them outside of a text editor.
- TypeScript checking is still in the core build. It can be turned off to save cpu/ram/time with `SKIP_TS_CHECK` (environment variable), and I've added an entry for how to run it as a separate step.

### Cool CI stuff

- I'm pretty sure I fixed (intermittent CI bug) #4337 for good.
- node_modules are cached so CI is usually faster.
    - This could also be applied to the Pytest workflows to make them a few minutes faster, too.

## Details (by commit)

### Speed improvements

- [x] [Speedup dev and prod builds with swc](https://github.com/kobotoolbox/kpi/pull/4323/commits/8d5a6ff0527f6e2cb8e1cdff282365b924fc706e)
    - Shared
        - Use `swc` as a faster drop-in replacement for babel
    - Dev
        - Use `react-refresh`, the successor to react-hot-loader
            - Certain small changes now update react components in-place
            - A full page reload is still required in other cases
    - Prod
        - Use `swc` when minifying
    - Tests
        - Ignore `['__proto__']` when comparing to mock objects
    - Note: This commit removes the TypeScript checking from the core build. We add it back later.
- [x] [Move slow linters out of shared build (run them separately)](https://github.com/kobotoolbox/kpi/pull/4323/commits/2fda342bd6db0dea5094ab04efde759fd2be35e8)
    - Significantly reduces start/build time
    - Where needed, linters can be run in a separate step:
        - in text editors - see kpi.code-workspace, #4230
        - as a separate cli step - see package.json "scripts"
    - Adds `lint-coffee` and `lint-styles` to package.json scripts
    - Note: This commit removes the ESLint step from CI, but it's added back later.
- [x] [Skip generate-icons step on dev server startup](https://github.com/kobotoolbox/kpi/pull/4323/commits/a0b4a91bd27250429a12b1293b6b670109ef5a8f)
    -  On `npm run watch`, skip `generate-icons`, because it…
        - seems to break cache / cause more work for webpack (by creating new files to process)
        - already runs on npm install
        - is only needed when `jsapp/svg-icons` have changes
    - Added a note to explain this instead (runs on `npm run watch`)
- [x] [Speedup storybook build](https://github.com/kobotoolbox/kpi/pull/4323/commits/19e5a5ff0c4360e336c84c5ffb9a98d2079ff78c)
    - Small improvement - more might be possible with upgrades to React or Storybook 7
        - Removed a slow plugin
        - Prod: Use `swc` when minifying.
- [x] [Toggleable TypeScript checking in webpack build](https://github.com/kobotoolbox/kpi/pull/4323/commits/321e5127d214a6d365bce30611430e031510427b)
    - Add TypeScript checking back into the webpack build/watch.
        - Skippable with `SKIP_TS_CHECK` to save cpu/ram/time
        - Prominent warning when `SKIP_TS_CHECK` is set, to make sure it's intentional
    - Add `check-types` in package.json for one-off TypeScript checks.
    - Misc
        - Also fix `something..gif` -> `something.gif`

### Other

- [x] [Allow running js tests in a browser tab](https://github.com/kobotoolbox/kpi/pull/4323/commits/ed94aed2112aefc03ac359af9cda647a06f15f5b) (useful to debug failed tests)
- [x] [Log detailed build measurements to console](https://github.com/kobotoolbox/kpi/pull/4323/commits/141bfe23e0512ad004f2098b42cb2822aa17fff6) (when `MEASURE_WEBPACK_PLUGIN_SPEED` is set)
- [x] [Remove packages no longer used](https://github.com/kobotoolbox/kpi/pull/4323/commits/0720a858d30604ef8c65a043a2a428ee8fc744b2) (and stop shipping an IE11 polyfill)
- [x] [CI: Try to mitigate occasional 'No inspectable targets' failures on npm test](https://github.com/kobotoolbox/kpi/pull/4323/commits/ffdd3d8416e8421592c0d248f7896a47f60fe487)
     - This 1st attempt [didn't fix the problem](https://github.com/kobotoolbox/kpi/actions/runs/4406745926/jobs/7719419067), but the next commit probably did.
- [x] [Run ESLint in CI, set a timeout for mocha-chrome, other misc changes.](https://github.com/kobotoolbox/kpi/pull/4323/commits/9a8e187403e316ea40764b2ad737a5295c8330ec)
    - Adds ESLint into CI (after build and TS check)
    - Labels CI steps
        - Install JavaScript dependencies (npm install)
        - Build Prod
        - Check TypeScript
        - Check ESLint, errors only
        - Build tests
        - Run tests, with mocha-chrome (timeout: 1 minute)
    - 2nd attempt for `#4337` - extends polling interval to 5s instead of adding more retries (which might be capped on some level)
- [x] [Remove deprecation warning on startup](https://github.com/kobotoolbox/kpi/pull/4323/commits/0dcffed9d419aa835e592d7fe1e15114e3c3da5c)
    - This patches `webpack-extract-translation-keys-plugin` (contributed [upstream](https://github.com/grassator/webpack-extract-translation-keys/pull/23))
- [x] [CI: Use cache for node_modules](https://github.com/kobotoolbox/kpi/pull/4323/commits/6e0885e29052bed19f9d8abcab3d356f353b4920) - in `npm-test`. (could add this to pytest actions too)
